scripts/stagingUtils.js

```diff
@@ -0,0 +1,308 @@
+        .then((result) => {
+          resolve(result.stdout.replace('\n', ''));
+        })
+        .catch((error) => {
+          console.error(`exec error: ${error}`);
+          reject(error);
+        });
+    });
+  },
+
+  async getGitUser() {
+    return new Promise((resolve, reject) => {
+      exec('git config --global user.name')
+        .then((result) => {
+          resolve(result.stdout.replace('\n', ''));
+        })
+        .catch((error) => {
+          console.error(`exec error: ${error}`);
+          reject(error);
+        });
+    });
+  },
+
+  async getGitCommits() {
+    const stdout = await exec('git cherry');
+    const cleanedup = stdout.replace(/\+ /g, '');
+    const commitarray = cleanedup.split(/\r\n|\r|\n/);
+    commitarray.pop(); // remove the last, dummy element that results from splitting on newline
+    if (commitarray.length === 0) {
+      console.error(
+        'You have tried to create a staging job from local commits but you have no committed work. Please make commits and then try again'
+      );
+      process.exit();
+    }
+    if (commitarray.length === 1) {
+      const firstCommit = commitarray[0];
+      const lastCommit = null;
+      return { firstCommit, lastCommit };
+    }
+    const firstCommit = commitarray[0];
+    const lastCommit = commitarray[commitarray.length - 1];
+    return { firstCommit, lastCommit };
+  },
+
+  getUpstreamName(upstream) {
+    console.log("yoooo ", upstream)
+    const upstreamInd = upstream.indexOf('origin/');
+    if (upstreamInd === -1) {
+      return upstream;
+    }
+    return 'master';
+  },
+
+  async checkUpstreamConfiguration(branchName) {
+
+    try {
+      const result = await exec(
+        `git rev-parse --abbrev-ref --symbolic-full-name ${branchName}@{upstream}`
+      );
+      return result.stdout;
+    } catch (error) {
+      if (error.code === 128) {
+        const errormsg = "You have not set an upstream for your local branch. Please do so with this command: \
+          \n\n \
+          git branch -u <upstream-branch-name>\
+          \n\n";
+        console.error(errormsg);
+        return errormsg;
+      }
+      console.error(error);
+      return error;
+    }
+  },
+
+  async doesRemoteHaveLocalBranch(branchName) {
+    try {
+      await exec(`git diff ${branchName} remotes/origin/${branchName}`);
+      return true;
+    } catch (error) {
+      if (error.code === 128) {
+        // we dont want to cancel the program
+        return false;
+      }
+      console.error(error);
+      return false;
+    }
+  },
+
+  async getGitPatchFromLocal(upstreamBranchName) {
+    return new Promise((resolve, reject) => {
+      exec(`git diff ${upstreamBranchName} --ignore-submodules > myPatch.patch`)
+        .then(() => {
+          fs.readFile('myPatch.patch', 'utf8', (err, data) => {
+              if (err) {
+                console.log("error reading patch file: ", err);
+                reject(err);
```

Comment:
will we get a `unhandledPromiseRejectionWarning` by rejecting within a `then` block? I could use some help here identifying the correct way to handle rejection